Bluetooth: Classic: add page/inquiry scan params and cod feature

chengkai15 opened the pull request:
This PR adds support for configuring BR/EDR page scan and inquiry scan parameters, and Class of Device (COD) features.

## Changes
1. **Page Scan Parameters**: 
   - Add `bt_br_page_scan_update_param` API to update page scan interval, window, and type (standard or interlaced).
   - Predefined parameter sets for different Bluetooth versions (R0, R1, R2) are provided.

2. **Inquiry Scan Parameters**:
   - Add `bt_br_inquiry_scan_update_param` API to update inquiry scan interval, window, and type.
   - Default parameter set is provided.

3. **Class of Device (COD)**:
   - Add APIs to set and get COD from the controller.

4. **Shell Commands**:
   - Add `pscan-param` shell command for testing page scan parameters.
   - Add `iscan-param` shell command for testing inquiry scan parameters.

## Testing
- Tested using the Bluetooth shell with various parameter sets and custom parameters.
- Verified that the page scan and inquiry scan parameters are correctly set in the controller.
- Verified COD set and get functionality.

## Notes
These features are essential for optimizing Bluetooth classic device discoverability and connectivity, and for correctly advertising the device class.

jhedberg commented on include/zephyr/bluetooth/classic/classic.h:
```diff
@@ -196,13 +196,79 @@ int bt_br_set_discoverable(bool enable, bool limited);
+ * @return Negative if fail set to requested state or requested state has been
+ *         already set. Zero if done successfully.
+ */
+int bt_br_write_page_scan_type(uint8_t type);
```
IMO these are too low level for the host stack API. They're essentially just 1:1 mapping to HCI. Please consider how we can abstract these in some nicer ways and if there are existing (possibly generic to both LE and BR/EDR) APIs that could be used instead.

jhedberg replied:
We could potentially have a single API to set all three parameters, with some config struct as input, and then have helper macros to construct those standard "Scenario" combinations from that table. That would retain the high level friendliness, but still allow someone to go and do some custom stuff if they really need it.

gzh-terry commented on include/zephyr/bluetooth/classic/classic.h:
```diff
@@ -196,13 +196,79 @@ int bt_br_set_discoverable(bool enable, bool limited);
+* @return Negative if fail set to requested state or requested state has been
+*         already set. Zero if done successfully.
+*/
+nt bt_br_set_connectable(bool enable);
```
`nt` -> `int`

gzh-terry commented on include/zephyr/bluetooth/hci_types.h:
```diff
@@ -635,6 +635,13 @@ struct bt_hci_rp_write_conn_accept_timeout {
+#define BT_HCI_OP_WRITE_INQUIRY_SCAN_ACTIVITY   BT_OP(BT_OGF_BASEBAND, 0x001e) /* 0x0c1e */
+struct bt_hci_cp_write_scan_activity {
+	uint16_t interval;
+	uint16_t windown;
```
windown -> window

gzh-terry commented on include/zephyr/bluetooth/classic/classic.h:
```diff
@@ -196,13 +196,79 @@ int bt_br_set_discoverable(bool enable, bool limited);
+int bt_br_write_inquiry_scan_activity(uint16_t interval, uint16_t window);
+
+/**
+ * @brief Set the inquiry Scan Type configuration parameter of the local
```
Only use upper case for the initial letter.

chengkai15 replied:
updated

chengkai15 committed:
Bluetooth: Classic: add COD set/get feature

This commit implements Class of Device (COD) functionality for
Classic Bluetooth, supporting both setting and retrieving COD
from the controller by bt_br_write_cod and bt_br_read_cod.

Signed-off-by: Kai Cheng <chengkai@xiaomi.com>

chengkai15 committed:
Bluetooth: Shell: add COD set/get support test shell

This commit implements Class of Device (COD) test shell
functionality for Classic Bluetooth, supporting both
setting and retrieving COD from the controller.

Signed-off-by: Kai Cheng <chengkai@xiaomi.com>

chengkai15 committed:
Bluetooth: Classic: add page scan parameters update

Implement BR/EDR page scan parameters update functionality with new
API bt_br_page_scan_update_param(). Adds support for setting page scan
interval, window and type (standard/interlaced) via HCI commands.
Includes predefined parameter sets for different Bluetooth versions
and use cases.

Signed-off-by: Kai Cheng <chengkai@xiaomi.com>

chengkai15 committed:
Bluetooth: Shell: add page scan parameters command

Add shell command for testing BR/EDR page scan parameters update.
Supports predefined parameter sets (R0, R1, R2 variants) and custom
parameters via interval/window/type arguments.

Signed-off-by: Kai Cheng <chengkai@xiaomi.com>

chengkai15 committed:
Bluetooth: Classic: add inquiry scan parameters update

Implement BR/EDR inquiry scan parameters update functionality with
new API bt_br_inquiry_scan_update_param(). Adds support for setting
inquiry scan interval, window and type via HCI commands. Includes
default parameter set for standard inquiry scanning.

Signed-off-by: Kai Cheng <chengkai@xiaomi.com>

chengkai15 committed:
Bluetooth: Shell: add inquiry scan parameters command

Add shell command for testing BR/EDR inquiry scan parameters update.
Supports default parameter set and custom parameters via interval/
window/type arguments for inquiry scan configuration.

Signed-off-by: Kai Cheng <chengkai@xiaomi.com>

jhedberg reviewed (changes requested):
@chengkai15 there's a "t" missing from the title of your second commit. Also, I'd recommend to familiarize yourself with https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-body and see if you think your commit messages need some improving :)

gzh-terry commented on include/zephyr/bluetooth/classic/classic.h:
```diff
@@ -196,13 +196,262 @@ int bt_br_set_discoverable(bool enable, bool limited);
  * resets discoverability if was set.
+*
+* @param enable Value allowing/disallowing controller to be connectable.
+*
```
Align codes.

gzh-terry commented on include/zephyr/bluetooth/classic/classic.h:
```diff
@@ -196,13 +196,262 @@ int bt_br_set_discoverable(bool enable, bool limited);
+
+/**
+ * @name Defined BR Page Scan timers
+ * @{
```
Remove

gzh-terry reviewed (dismissed):
Generally looks fine to me, but pay attention to code styles.

lylezhu2012 commented on subsys/bluetooth/host/classic/br.c:
```diff
@@ -1296,6 +1296,40 @@ int bt_br_set_discoverable(bool enable, bool limited)
+		return -EAGAIN;
+	}
+
+	err = bt_br_read_cod(cod);
```
I think the param `cod` needs to be checked before using it.

lylezhu2012 commented on subsys/bluetooth/host/classic/br.c:
```diff
@@ -1296,6 +1296,40 @@ int bt_br_set_discoverable(bool enable, bool limited)
+		return -EAGAIN;
+	}
+
+	err = bt_br_write_cod(cod);
```
I believe the limited discover mode bit should not be cleared if it is in limited discovery mode when updating the COD.

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -1746,6 +1791,7 @@ SHELL_STATIC_SUBCMD_SET_CREATE(br_cmds,
 		      cmd_discovery, 2, 2),
 	SHELL_CMD_ARG(iscan, NULL, "<value: on, off> [mode: limited]",
 		      cmd_discoverable, 2, 1),
+	SHELL_CMD_ARG(iscan-param, NULL, HELP_ISCAN_PARAM, cmd_iscan_param, 0, 3),
```
```suggestion
	SHELL_CMD_ARG(iscan-param, NULL, HELP_ISCAN_PARAM, cmd_iscan_param, 1, 3),
```

chengkai15 commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -1469,6 +1469,38 @@ static int cmd_set_role_switchable(const struct shell *sh, size_t argc, char *ar
+			shell_error(sh, "fail to set cod (err %d)", err);
+		} else {
+			shell_print(sh, "set cod:0x%04x success", cod);
+		}
```
updated

chengkai15 commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -1614,6 +1691,12 @@ static int cmd_default_handler(const struct shell *sh, size_t argc, char **argv)
+	"[mode: r0, fr1, mr1, sr1, fr2, sr2]\n"                                                    \
+	"[<interval: scan interval in units of 0.625 ms>\n"                                        \
+	" <window: window in units of 0.625 ms>\n"                                                 \
+	" <type: 0 for standard, 1 for interlaced>]"
```
divided into cmd_pscan_mode and cmd_pscan_param part

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -1730,6 +1730,38 @@ static int cmd_set_sniff_mode(const struct shell *sh, size_t argc, char *argv[])
+	uint32_t cod;
+
+	err = bt_br_get_class_of_device(&cod);
+	if (err) {
```
```suggestion
	if (err != 0) {
```

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -1730,6 +1730,38 @@ static int cmd_set_sniff_mode(const struct shell *sh, size_t argc, char *argv[])
+	cod = strtoul(argv[1], NULL, 16);
+
+	err = bt_br_set_class_of_device(cod);
+	if (err) {
```
```suggestion
	if (err != 0) {
```

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -1730,6 +1730,38 @@ static int cmd_set_sniff_mode(const struct shell *sh, size_t argc, char *argv[])
+	if (err) {
+		shell_error(sh, "fail to set cod (err %d)", err);
+	} else {
+		shell_print(sh, "set cod:0x%04x success", cod);
```
The size of cod should be 3 octets.
```suggestion
		shell_print(sh, "set cod:0x%06x success", cod);
```

lylezhu2012 commented on include/zephyr/bluetooth/classic/classic.h:
```diff
@@ -204,6 +204,188 @@ int bt_br_set_discoverable(bool enable, bool limited);
+
+/**
+ * @brief Get the Class of Device configuration parameter of the local
+ *        BR/EDR Controller.
```
```suggestion
 * @brief Get the Class of Device of the local BR/EDR Controller.
```

lylezhu2012 commented on subsys/bluetooth/host/classic/br.c:
```diff
@@ -1330,6 +1334,55 @@ int bt_br_set_discoverable(bool enable, bool limited)
+	}
+
+	err = bt_br_write_cod(cod);
+	if (err) {
```
```suggestion
	if (err != 0) {
```

lylezhu2012 commented on subsys/bluetooth/host/classic/br.c:
```diff
@@ -1334,6 +1334,90 @@ int bt_br_set_discoverable(bool enable, bool limited)
+	}
+
+	buf = bt_hci_cmd_alloc(K_FOREVER);
+	if (!buf) {
```
```suggestion
	if (buf == NULL) {
```

lylezhu2012 commented on subsys/bluetooth/host/classic/br.c:
```diff
@@ -1334,6 +1334,90 @@ int bt_br_set_discoverable(bool enable, bool limited)
+		return err;
+	}
+
+	return 0;
```
```suggestion
	return bt_hci_cmd_send_sync(opcode, buf, NULL);
```

chengkai15 replied:
updated

lylezhu2012 commented on subsys/bluetooth/host/classic/br.c:
```diff
@@ -1334,6 +1334,90 @@ int bt_br_set_discoverable(bool enable, bool limited)
+
+	err = bt_br_write_scan_activity(BT_HCI_OP_WRITE_PAGE_SCAN_ACTIVITY, param->interval,
+					param->window);
+	if (err) {
```
```suggestion
	if (err != 0) {
```

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -818,6 +818,99 @@ static int cmd_connectable(const struct shell *sh, size_t argc, char *argv[])
+	if (!strcmp(argv[1], "r0")) {
+		err = bt_br_page_scan_update_param(BT_BR_PAGE_SCAN_PARAM_R0);
+		if (err) {
+			shell_print(sh, "BR/EDR update page scan param failed (err %d)", err);
```
```suggestion
			shell_error(sh, "BR/EDR update page scan param failed (err %d)", err);
```

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -818,6 +818,99 @@ static int cmd_connectable(const struct shell *sh, size_t argc, char *argv[])
+		shell_print(sh, "BR/EDR update page scan param R0 success");
+	} else if (!strcmp(argv[1], "fr1")) {
+		err = bt_br_page_scan_update_param(BT_BR_PAGE_SCAN_PARAM_FAST_R1);
+		if (err) {
```
```suggestion
		if (err != 0) {
```

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -818,6 +818,99 @@ static int cmd_connectable(const struct shell *sh, size_t argc, char *argv[])
+	} else {
+		shell_help(sh);
+		return SHELL_CMD_HELP_PRINTED;
+	}
```
```suggestion
	struct bt_br_page_scan_param *param;

	if (!strcmp(argv[1], "r0")) {
		param = BT_BR_PAGE_SCAN_PARAM_R0;
	} else if (!strcmp(argv[1], "fr1")) {
		param = BT_BR_PAGE_SCAN_PARAM_FAST_R1;
	} else if (!strcmp(argv[1], "mr1")) {
		param = BT_BR_PAGE_SCAN_PARAM_MEDIUM_R1;
	} else if (!strcmp(argv[1], "sr1")) {
		param = BT_BR_PAGE_SCAN_PARAM_SLOW_R1;
	} else if (!strcmp(argv[1], "fr2")) {
		param = BT_BR_PAGE_SCAN_PARAM_FAST_R2;
	} else if (!strcmp(argv[1], "sr2")) {
		param = BT_BR_PAGE_SCAN_PARAM_SLOW_R2;
	} else {
		shell_help(sh);
		return SHELL_CMD_HELP_PRINTED;
	}

	err = bt_br_page_scan_update_param(param);
	if (err != 0) {
		shell_error(sh, "BR/EDR update page scan param failed (err %d)", err);
		return -ENOEXEC;
	}

	shell_print(sh, "BR/EDR update page scan param %s success", argv[1]);
```

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -818,6 +818,99 @@ static int cmd_connectable(const struct shell *sh, size_t argc, char *argv[])
+	if (err) {
+		shell_print(sh, "BR/EDR update page scan param failed (err %d)", err);
+		return -ENOEXEC;
+	}
```
```suggestion
	if (err != 0) {
		shell_error(sh, "BR/EDR update page scan param failed (err %d)", err);
		return -ENOEXEC;
	}
```

chengkai15 replied:
updated

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -792,6 +792,48 @@ static int cmd_discoverable(const struct shell *sh, size_t argc, char *argv[])
+			    BT_BR_INQUIRY_SCAN_PARAM_DEFAULT->window,
+			    BT_BR_INQUIRY_SCAN_PARAM_DEFAULT->type);
+	} else if (argc == 4) {
+		struct bt_br_inquiry_scan_param param = {0};
```
```suggestion
		struct bt_br_inquiry_scan_param param;
```

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -792,6 +792,48 @@ static int cmd_discoverable(const struct shell *sh, size_t argc, char *argv[])
+			    "type:%u) success",
+			    BT_BR_INQUIRY_SCAN_PARAM_DEFAULT->interval,
+			    BT_BR_INQUIRY_SCAN_PARAM_DEFAULT->window,
+			    BT_BR_INQUIRY_SCAN_PARAM_DEFAULT->type);
```
```suggestion
		shell_print(sh,
			    "BR/EDR update inquiry scan param(interval:0x%04x, window:0x%04x, "
			    "type:%u) success",
			    BT_BR_INQUIRY_SCAN_PARAM_DEFAULT->interval,
			    BT_BR_INQUIRY_SCAN_PARAM_DEFAULT->window,
			    BT_BR_INQUIRY_SCAN_PARAM_DEFAULT->type);
```

jhedberg replied:
> interval and window are uint16_t type, then it need 0x%02x print type

No. The given number is how many characters are used. `uint8_t` should have `%02x` and `uint16_t` should have `%04x`.

chengkai15 replied:
got it. updated

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -792,6 +792,48 @@ static int cmd_discoverable(const struct shell *sh, size_t argc, char *argv[])
+		shell_print(sh,
+			    "BR/EDR update inquiry scan param(interval:0x%02x, window:0x%02x, "
+			    "type:%u) success",
+			    param.interval, param.window, param.type);
```
```suggestion
		shell_print(sh,
			    "BR/EDR update inquiry scan param(interval:0x%04x, window:0x%04x, "
			    "type:%u) success",
			    param.interval, param.window, param.type);
```

chengkai15 replied:
interval and window are uint16_t type, then it need 0x%02x print type

jhedberg replied:
No. The number is about printed characters. `uint16_t` needs four.

chengkai15 commented on include/zephyr/bluetooth/classic/classic.h:
```diff
@@ -204,6 +204,188 @@ int bt_br_set_discoverable(bool enable, bool limited);
+#define BT_COD_MINOR_DEVICE_CLASS_KNEE_PROSTHESIS           0x0C
+#define BT_COD_MINOR_DEVICE_CLASS_ANKLE_PROSTHESIS          0x0D
+#define BT_COD_MINOR_DEVICE_CLASS_GENERIC_HEALTH_MANAGER    0x0E
+#define BT_COD_MINOR_DEVICE_CLASS_PERSONAL_MOBILITY_DEVICE  0x0F
```
updated

chengkai15 commented on subsys/bluetooth/host/classic/br.c:
```diff
@@ -1330,6 +1334,55 @@ int bt_br_set_discoverable(bool enable, bool limited)
+
+	err = bt_br_write_cod(cod);
+	if (err) {
+		LOG_ERR("write cod:0x%08x failed (err %d)", cod, err);
```
updated

chengkai15 commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -818,6 +818,99 @@ static int cmd_connectable(const struct shell *sh, size_t argc, char *argv[])
+	if (argc != 2) {
+		shell_help(sh);
+		return SHELL_CMD_HELP_PRINTED;
+	}
```
updated

chengkai15 commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -818,6 +818,99 @@ static int cmd_connectable(const struct shell *sh, size_t argc, char *argv[])
+	if (argc != 4) {
+		shell_help(sh);
+		return SHELL_CMD_HELP_PRINTED;
+	}
```
updated

chengkai15 commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -1973,6 +2015,11 @@ static int cmd_default_handler(const struct shell *sh, size_t argc, char **argv)
 	"<type: 0 for standard, 1 for interlaced>"
 
+#define HELP_ISCAN_PARAM                                                                           \
+	"<interval: scan interval in units of 0.625 ms>\n"                                         \
```
updated

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -818,6 +818,61 @@ static int cmd_connectable(const struct shell *sh, size_t argc, char *argv[])
+static int cmd_pscan_param(const struct shell *sh, size_t argc, char *argv[])
+{
+	int err;
+	struct bt_br_page_scan_param param = {0};
```
```suggestion
	struct bt_br_page_scan_param param;
```

lylezhu2012 commented on include/zephyr/bluetooth/classic/classic.h:
```diff
@@ -204,6 +204,200 @@ int bt_br_set_discoverable(bool enable, bool limited);
+#define BT_BR_PAGE_SCAN_FAST_WINDOW_R2   0x0011 /* 0x0011, 10.625ms, U:0.625 */
+
+#define BT_BR_PAGE_SCAN_SLOW_INTERVAL_R2 0x1000 /* 0x1000, 2.560s, U:0.625 */
+#define BT_BR_PAGE_SCAN_SLOW_WINDOW_R2   0x0011 /* 0x0011, 10.625ms, U:0.625 */
```
Please use comments in doxygen format . For example, change `/* 0x0012, 11.25ms, U:0.625 */` to `/** 0x0012, 11.25ms, U:0.625 */`.

lylezhu2012 commented on include/zephyr/bluetooth/classic/classic.h:
```diff
@@ -398,6 +398,92 @@ struct bt_br_page_scan_param {
+ */
+
+#define BT_BR_INQUIRY_SCAN_INTERVAL_DEFAULT 0x1000 /* 0x1000, 2.560s, U:0.625 */
+#define BT_BR_INQUIRY_SCAN_WINDOW_DEFAULT   0x0012 /* 0x0012, 11.25mss, U:0.625 */
```
Please change `/* ... */` to `/** ... */`.

chengkai15 commented on include/zephyr/bluetooth/classic/classic.h:
```diff
@@ -204,6 +204,200 @@ int bt_br_set_discoverable(bool enable, bool limited);
+}
+
+/**
+ * Helper to declare BR/EDR page scan parameters inline
```
updated

chengkai15 commented on include/zephyr/bluetooth/classic/classic.h:
```diff
@@ -204,6 +204,200 @@ int bt_br_set_discoverable(bool enable, bool limited);
+ *
+ * @param _interval page scan interval, N * 0.625 milliseconds
+ * @param _window   page scan window, N * 0.625 milliseconds
+ * @param _type    BT_BR_SCAN_TYPE_STANDARD or BT_BR_SCAN_TYPE_INTERLACED
```
updated

chengkai15 commented on include/zephyr/bluetooth/hci_types.h:
```diff
@@ -712,6 +712,12 @@ struct bt_hci_rp_write_conn_accept_timeout {
 #define BT_BREDR_SCAN_PAGE                      0x02
 
+#define BT_HCI_OP_WRITE_PAGE_SCAN_ACTIVITY      BT_OP(BT_OGF_BASEBAND, 0x001c) /* 0x0c1c */
+struct bt_hci_cp_write_scan_activity {
```
bt_hci_cp_write_scan_activity was not only for page scan, but also for inquiry scan as well.

so I suggest not rename it

lylezhu2012 replied:
I believe that the dedicated structure should be defined for each HCI command, rather than a generic structure shared by multiple HCI commands.

It has several advantages:
1. Easy to read,
2. The compiler can perform type checking,
3. Easy to maintain.

And I am not sure if other HCI commands have the similar cases. But I did not find it.

chengkai15 commented on include/zephyr/bluetooth/classic/classic.h:
```diff
@@ -398,6 +398,92 @@ struct bt_br_page_scan_param {
+ */
+
+/**
+ * Helper to declare BR/EDR inquiry scan parameters inline
```
updated

makeshi commented on include/zephyr/bluetooth/classic/classic.h:
```diff
@@ -204,6 +204,201 @@ int bt_br_set_discoverable(bool enable, bool limited);
+ */
+
+#define BT_BR_PAGE_SCAN_PARAM(_interval, _window, _type) \
+	((const struct bt_br_page_scan_param[]) { \
```
Does need static?

makeshi commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -1730,6 +1730,38 @@ static int cmd_set_sniff_mode(const struct shell *sh, size_t argc, char *argv[])
+		shell_error(sh, "fail to set cod (err %d)", err);
+	} else {
+		shell_print(sh, "set cod:0x%06x success", cod);
+	}
```
suggest change to : 
	if (err != 0) {
		shell_error(sh, "fail to set cod (err %d)", err);
		return err;
	}
	
	shell_print(sh, "get cod:0x%06x success", cod);

chengkai15 replied:
updated

makeshi commented on subsys/bluetooth/host/classic/br.c:
```diff
@@ -1334,6 +1334,84 @@ int bt_br_set_discoverable(bool enable, bool limited)
+	if (param->interval < param->window) {
+		return -EINVAL;
+	}
+
```
Does need check param->type?

chengkai15 replied:
updated

makeshi commented on subsys/bluetooth/host/classic/shell/bredr.c:
```diff
@@ -792,6 +792,50 @@ static int cmd_discoverable(const struct shell *sh, size_t argc, char *argv[])
+
+		err = bt_br_inquiry_scan_update_param(&param);
+		if (err != 0) {
+			shell_error(sh, "BR/EDR update inquiry scan param failed (err %d)", err);
```
It is better let the error message different line 802

lylezhu2012 commented on include/zephyr/bluetooth/classic/classic.h:
```diff
@@ -204,6 +204,325 @@ int bt_br_set_discoverable(bool enable, bool limited);
+ *
+ * @param _interval page scan interval, N * 0.625 milliseconds
+ * @param _window   page scan window, N * 0.625 milliseconds
+ * @param _type     BT_BR_SCAN_TYPE_STANDARD or BT_BR_SCAN_TYPE_INTERLACED
```
```suggestion
 * @param _type     @ref BT_BR_SCAN_TYPE_STANDARD or @ref BT_BR_SCAN_TYPE_INTERLACED
```

chengkai15 commented on include/zephyr/bluetooth/classic/classic.h:
```diff
@@ -204,6 +204,325 @@ int bt_br_set_discoverable(bool enable, bool limited);
+ * @brief Fast page scan parameters for R1
+ *
+ * Page scan interval is set to 100 ms (0x00A0 in 0.625 ms units), and the
+ * page scan window is set to 10.240 seconds (0x27FF in 1 ms units).
```
updated

lylezhu2012 reviewed (commented):
It is OK for me

lylezhu2012 commented on include/zephyr/bluetooth/assigned_numbers.h:
```diff
@@ -1462,6 +1462,155 @@ enum bt_audio_location {
+#define BT_COD_MINOR_DEVICE_CLASS_KNEE_PROSTHESIS           0x0C
+#define BT_COD_MINOR_DEVICE_CLASS_ANKLE_PROSTHESIS          0x0D
+#define BT_COD_MINOR_DEVICE_CLASS_GENERIC_HEALTH_MANAGER    0x0E
+#define BT_COD_MINOR_DEVICE_CLASS_PERSONAL_MOBILITY_DEVICE  0x0F
```
Please also fix the doxygen comments issue.
